fix(common): ensure scrollRestoration is writable

marcvincenti opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #30629


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented:
Hi @mvrcrypto 
Thanks for sending in this Pull Request.
Before we can review and potentially merge it there are various things that need fixing.

1) Please sign the CLA
2) We cannot accept a change for a bug unless there are unit tests to demonstrate the bug and that it is fixed. There is a basic unit test in the other PR that you linked, which could act as a starting point: https://github.com/taion/scroll-behavior/pull/127/files
3) Our commit messages need to be in a specific format so that we can generate change log information from them, and also to provide information for developers looking at the changes in the future.

petebacondarwin reviewed (changes requested):
I wonder if we could test this "settability" in the `supportScrollRestoraton()` function and when defining `supportManualScrollRestoration` instead?

mvrcrypto commented:
I signed it!

mvrcrypto commented:
Ok, i will add some tests and update commit messages.
Also i agree with you on testing the settability but i'm not sure how to do it and if this method is called a lot. If yes, we should definitely set a property `supportManualScrollRestoration`.

petebacondarwin commented:
Perhaps something along the lines of (in `viewport_scroller.ts`):

```ts

  /**
   * We only support scroll restoration when we can get a hold of window.
   * This means that we do not support this behavior when running in a web worker.
   *
   * Lifting this restriction right now would require more changes in the dom adapter.
   * Since webworkers aren't widely used, we will lift it once RouterScroller is
   * battle-tested.
   */
  private supportScrollRestoration(): boolean {
    try {
      // check that `scrollRestoration` is not just a getter.
      this.window.history.scrollRestoration = this.window.history.scrollRestoration;
      return !!this.window && !!this.window.scrollTo;
    } catch {
      return false;
    }
  }
```

petebacondarwin commented on aio/src/app/shared/scroll.service.ts:
```diff
@@ -56,7 +56,11 @@ export class ScrollService {
+        history.scrollRestoration = 'manual';
+      } catch (e) {
+        this.supportManualScrollRestoration = false;
+      }
```
And now something similar here... change the place that we define `supportManualScrollRestoration` to do the try-catch rather than here.

alfaproject commented on packages/common/src/viewport_scroller.ts:
```diff
@@ -164,6 +164,8 @@ export class BrowserViewportScroller implements ViewportScroller {
   private supportScrollRestoration(): boolean {
     try {
+      // check that `scrollRestoration` is not just a getter.
+      this.window.history.scrollRestoration = this.window.history.scrollRestoration;
```
Would't using `getOwnPropertyDescriptor` be more clean?

petebacondarwin replied:
Not a bad idea. I was thinking that catching the exception covers cases were it is not writable but is not necessarily diagnosable via the property descriptor, like if it is an internal thing rather than configured in JS.

petebacondarwin replied:
We could try it and see if we stop getting these error reports.

petebacondarwin replied:
It turns out that `window.history` does not own the `scrollRestoration` property. It is on its prototype.

petebacondarwin replied:
Also it is not `writable`!! Instead it has a getter and a setter!

Commit:
fix(common): ensure scrollRestoration is writable

Some specialised browsers that do not support scroll restoration
(e.g. some web crawlers) do not allow `scrollRestoration` to be
writable.

We already sniff the browser to see if it has the `window.scrollTo`
method, so now we also check whether `window.history.scrollRestoration`
is writable too.

Fixes #30629

petebacondarwin commented on aio/src/app/shared/scroll.service.ts:
```diff
@@ -24,7 +24,9 @@ export class ScrollService {
-      !!history && ('scrollRestoration' in history);
+      !!history && ('scrollRestoration' in history) &&
+      !!Object.getOwnPropertyDescriptor(history, 'scrollRestoration') &&
+      !!Object.getOwnPropertyDescriptor(history, 'scrollRestoration')!.writable;
```
I think this will need some formatting.

petebacondarwin replied:
Try running `yarn gulp format`

petebacondarwin reviewed (approved):
Loving the tests! Thanks @mvrcrypto 

Check the formatting and then when CI is happy I will remove the `cleanup` tag.

alfaproject commented on packages/common/src/viewport_scroller.ts:
```diff
@@ -164,7 +164,9 @@ export class BrowserViewportScroller implements ViewportScroller {
-      return !!this.window && !!this.window.scrollTo;
+      return !!this.window && !!this.window.scrollTo &&
+        !!Object.getOwnPropertyDescriptor(history, 'scrollRestoration') &&
+        !!Object.getOwnPropertyDescriptor(history, 'scrollRestoration')!.writable;
```
Accessing `history` at top level doesn't seem SSR safe to me, maybe rewrite this:
```ts
      if (!this.window || !this.window.scrollTo) {
        return false;
      }

      const scrollRestorationDescriptor = Object.getOwnPropertyDescriptor(this.window.history, 'scrollRestoration');
      return !!scrollRestorationDescriptor && scrollRestorationDescriptor.writable;
```

petebacondarwin replied:
Oops good spot @alfaproject - thanks again!

petebacondarwin replied:
Saves on the `!` operator too!

mvrcrypto commented:
Thank you for your help @petebacondarwin. I just have one question regarding the tests. I'm not able to run the *viewport_scroller* tests. I get `The test command requires to be run in an Angular project, but a project definition could not be found.` What does it means ? Can i run it simply with the aio tests ?

petebacondarwin commented:
You need to use Bazel to run the ViewPort tests:

```bash
bazel test //packages/common/test
```

petebacondarwin commented:
I rebased and tweaked the common tests. I think the AIO tests are still not happy.
Do you want to take another look at this @mvrcrypto ?

petebacondarwin committed:
style(docs-infra): reformat ScrollService file

Pre-empting code formatting changes when the
code is updated in a subsequent commit.

petebacondarwin commented:
I've cleaned this up and got the tests passing.

jelbourn commented:
@petebacondarwin does 751 bytes seem reasonable for this change?

petebacondarwin commented:
> @petebacondarwin does 751 bytes seem reasonable for this change?

Actually the change is only 360 bytes. I just updated the PR and it is now 429944 rather than 429951. Moreover `master` is not represented accurately in the goldens file. It has obviously crept up in the background because master is currently 429584, rather than 429200.

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

atscott reviewed (approved):
Reviewed-for: size-tracking

mhevery reviewed (approved):
reviewed-for: fw-core

mhevery commented:
[presubmit](http://test/OCL:324096928:BASE:324096984:1596148409945:5610ae13)

AndrewKushnir commented:
[Presubmit](http://test/OCL:325089517:BASE:325089542:1596659431196:a8d6ceb3)
